### ISSUE TITLE
dbus-mpris only works after spotify-protocol comms

### ISSUE THREAD

**Issue by wasperen** [Spotifyd/spotifyd#185]
Thanks for adding dbus communitation! I got it working. Quite straight forward.
The only thing I found is that I do first need to talk to this daemon through the Spotify connect protocol. After that is done once, I can talk mpris to it.

**Comment by wasperen** [comment]
By the way: this is when running as `no-daemon`

**Comment by stsydow** [comment]
I have the same problem.
The command seems to arrive at spotifyd but nothing happens.

Here is my trace:
```
...
17:10:08 [DEBUG] dbus_tokio::adriver: handle_msgs: (MethodCall, Some("/org/mpris/MediaPlayer2"), Some("org.freedesktop.DBus.Properties"), Some("GetAll"))
17:10:08 [DEBUG] tokio_core::reactor: loop poll - 374.083µs
17:10:08 [DEBUG] tokio_core::reactor: loop time - Instant { tv_sec: 25101, tv_nsec: 694219180 }
17:10:08 [DEBUG] tokio_core::reactor: loop process, 9.863µs
17:10:08 [DEBUG] dbus_tokio::adriver: D-Bus i/o poll ready: 23 is Ready(Readable)
17:10:08 [DEBUG] dbus_tokio::adriver: D-Bus i/o unix ready: 23 is Readable | Aio
17:10:08 [DEBUG] tokio_reactor::registration: scheduling Read for: 4
17:10:08 [DEBUG] tokio_core::reactor: loop poll - 46.837µs
17:10:08 [DEBUG] tokio_core::reactor: loop time - Instant { tv_sec: 25101, tv_nsec: 694292496 }
17:10:08 [DEBUG] tokio_core::reactor: loop process, 10.384µs
17:10:21 [TRACE] tokio_reactor: [<unknown>:376] event Readable Token(25165828)
17:10:21 [TRACE] tokio_reactor: [<unknown>:390] loop process - 1 events, 0.000s
17:10:21 [TRACE] tokio_reactor: [<unknown>:376] event Readable Token(25165828)
17:10:21 [TRACE] tokio_reactor: [<unknown>:390] loop process - 1 events, 0.000s
17:10:21 [TRACE] tokio_reactor: [<unknown>:376] event Readable Token(25165828)
17:10:21 [TRACE] tokio_reactor: [<unknown>:390] loop process - 1 events, 0.000s
17:10:21 [TRACE] tokio_reactor: [<unknown>:376] event Readable Token(25165828)
17:10:21 [TRACE] tokio_reactor: [<unknown>:390] loop process - 1 events, 0.000s
17:10:21 [TRACE] tokio_reactor: [<unknown>:376] event Readable Token(25165828)
17:10:21 [TRACE] tokio_reactor: [<unknown>:390] loop process - 1 events, 0.000s
17:10:21 [DEBUG] dbus_tokio::adriver: D-Bus i/o poll ready: 23 is Ready(Readable)
17:10:21 [DEBUG] dbus_tokio::adriver: D-Bus i/o unix ready: 23 is Readable | Aio
17:10:21 [TRACE] tokio_reactor: [<unknown>:376] event Readable Token(25165828)
17:10:21 [TRACE] tokio_reactor: [<unknown>:390] loop process - 1 events, 0.000s
17:10:21 [DEBUG] dbus_tokio::adriver: handle_msgs: (MethodCall, Some("/org/mpris/MediaPlayer2"), Some("org.mpris.MediaPlayer2.Player"), Some("PlayPause"))
17:10:21 [DEBUG] tokio_core::reactor: loop poll - 12.99865996s
17:10:21 [DEBUG] tokio_core::reactor: loop time - Instant { tv_sec: 25114, tv_nsec: 692973752 }
17:10:21 [DEBUG] tokio_core::reactor: loop process, 9.552µs
17:10:21 [DEBUG] dbus_tokio::adriver: D-Bus i/o poll ready: 23 is Ready(Readable)
17:10:21 [DEBUG] dbus_tokio::adriver: D-Bus i/o unix ready: 23 is Readable | Aio
17:10:21 [DEBUG] tokio_reactor::registration: scheduling Read for: 4
17:10:21 [DEBUG] tokio_core::reactor: loop poll - 18.652µs
17:10:21 [DEBUG] tokio_core::reactor: loop time - Instant { tv_sec: 25114, tv_nsec: 693008458 }
17:10:21 [DEBUG] tokio_core::reactor: loop process, 7.235µs
```
```
$ sp play
Error org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.
```

**Comment by KittyMac** [comment]
I also experienced this behavior.  I fixed it with the following code change:

https://github.com/KittyMac/spotifyd/commit/9ebf7730ac627dde657fccf3855cbc93e41dc823

**Comment by SirWindfield** [comment]
Oh wow, I just saw this, thanks! @KittyMac. I'll create a PR and merge it as soon as I can!

**Comment by SirWindfield** [comment]
Ok I think I misunderstood you before. If you start up the daemon, it just announces itself as a spotify connect client. Nothing more. You first need to give it a command to actively play music (aka switch to the spotifyd daemon as the audio output) so it can actually start playing songs. Unit you do that, the daemon doesn't know what songs you are playing right now.

Taking that into account, it makes sense that mpris doesn't work until you interact with the daemon beforehand through the Spotify Connect feature.

I hope this clarifies things.

**Comment by awerlang** [comment]
@SirWindfield 

> Ok I think I misunderstood you before. If you start up the daemon, it just announces itself as a spotify connect client. Nothing more. You first need to give it a command to actively play music (aka switch to the spotifyd daemon as the audio output) so it can actually start playing songs. Unit you do that, the daemon doesn't know what songs you are playing right now.
> 
> Taking that into account, it makes sense that mpris doesn't work until you interact with the daemon beforehand through the Spotify Connect feature.
> 
> I hope this clarifies things.


From https://github.com/Spotifyd/spotifyd#configuration

> Spotifyd is able to run without configuration at all and will assume default values for most of the fields. However, running without configuration will only allow you to connect to it via Spotify Connect if you're on the same network as the daemon.

The README explains that at least one option to use the tool exists that's not through Spotify Connect. I guess that's why there's username and password settings. Spotify Connect doesn't need these settings.

MPRIS can receive an OpenUri command, so it should be ready regardless of Spotify Connect.

If @KittyMac patch does in fact fix this issue, then the `else` can be removed, since both code paths end with the same `poll()`.

**Comment by awerlang** [comment]
I attempted @KittyMac approach, only with the change I proposed. It didn't work for me.

> Couldn't fetch metadata from spotify: Ok(None)
> Couldn't fetch metadata from spotify: Ok(None)

The first message I would always get even with the unmodified spotifyd. The second message is new (as in at the time an attempt to play is made), it seems the MPRIS interface is up, but doesn't attempt to load a track on command.

Then, after an hour is passed:

> Caught panic with message: cannot poll Map twice

and crashes.
